src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProjectFactory.cs

```diff
@@ -305,17 +329,21 @@ internal void RemoveSolution_NoLock()
             Contract.ThrowIfFalse(_projectToMaxSupportedLangVersionMap.Count == 0);
             Contract.ThrowIfFalse(_projectToDependencyNodeTargetIdentifier.Count == 0);
 
-            // Create a new empty solution and set this; we will reuse the same SolutionId and path since components still may have persistence information they still need
-            // to look up by that location; we also keep the existing analyzer references around since those are host-level analyzers that were loaded asynchronously.
-            Workspace.ClearOpenDocuments();
+            // Create a new empty solution and set this; we will reuse the same SolutionId and path since components
+            // still may have persistence information they still need to look up by that location; we also keep the
+            // existing analyzer references around since those are host-level analyzers that were loaded asynchronously.
 
             Workspace.SetCurrentSolution(
                 solution => Workspace.CreateSolution(
                     SolutionInfo.Create(
                         SolutionId.CreateNewId(),
                         VersionStamp.Create(),
                         analyzerReferences: solution.AnalyzerReferences)),
-                WorkspaceChangeKind.SolutionRemoved);
+                WorkspaceChangeKind.SolutionRemoved,
+                onBeforeUpdate: (_, _) =>
+                {
+                    Workspace.ClearOpenDocuments();
```

Comment:
an example of moving mutation to the point that we know that setting the actual solution succeeded.  It was previously possible for these two calls to have interleaving mutations run from some other source